src/cpp/src/rag/text_embedding_pipeline.cpp

```diff
@@ -0,0 +1,290 @@
+        m_request = compiled_model.create_infer_request();
+    };
+
+    std::vector<EmbeddingResult> embed_documents(const std::vector<std::string>& texts) {
+        start_embed_documents_async(texts);
+        return wait_embed_documents();
+    };
+
+    void start_embed_documents_async(const std::vector<std::string>& texts) {
+        auto formatted_texts = format_texts(texts);
+        start_embed_async(formatted_texts);
+    };
+
+    std::vector<EmbeddingResult> wait_embed_documents() {
+        return wait_embed();
+    };
+
+    EmbeddingResult embed_query(const std::string& text) {
+        start_embed_query_async(text);
+        return wait_embed_query();
+    };
+
+    void start_embed_query_async(const std::string& text) {
+        std::vector<std::string> formatted_query{format_query(text)};
+        start_embed_async(formatted_query);
+    };
+
+    EmbeddingResult wait_embed_query() {
+        return wait_embed()[0];
+    };
+
+private:
+    Tokenizer m_tokenizer;
+    InferRequest m_request;
+    Config m_config;
+    AnyMap m_tokenization_params;
+
+    void start_embed_async(std::vector<std::string>& texts) {
+        const auto encoded = m_tokenizer.encode(texts, m_tokenization_params);
+
+        m_request.set_tensor("input_ids", encoded.input_ids);
+        m_request.set_tensor("attention_mask", encoded.attention_mask);
+
+        // fill token_type_ids
+        // todo: pass token_type_ids from tokenizer
+        for (auto& input : m_request.get_compiled_model().inputs()) {
+            if (input.get_any_name() == "token_type_ids") {
+                ov::Tensor token_type_ids{ov::element::i64, encoded.input_ids.get_shape()};
+                std::fill_n(token_type_ids.data<int64_t>(), encoded.input_ids.get_size(), 0);
+                m_request.set_tensor("token_type_ids", token_type_ids);
+                break;
+            }
+        }
+
+        m_request.start_async();
+    };
+
+    std::vector<EmbeddingResult> wait_embed() {
+        m_request.wait();
+
+        // [batch_size, hidden_size]
+        const Tensor last_hidden_state = m_request.get_tensor("last_hidden_state");
+
+        return to_embedding_result(last_hidden_state);
+    };
+
+    std::vector<std::string> format_texts(const std::vector<std::string>& texts) {
+        if (!m_config.embed_instruction) {
+            return texts;
+        }
+
+        std::vector<std::string> formatted;
+        formatted.reserve(texts.size());
+
+        for (auto& text : texts) {
+            formatted.emplace_back(*m_config.embed_instruction + text);
+        }
+        return formatted;
+    }
+
+    std::string format_query(const std::string& text) {
+        if (!m_config.query_instruction) {
+            return text;
+        }
+
+        return *m_config.query_instruction + text;
+    }
+
+    std::vector<EmbeddingResult> to_embedding_result(const Tensor& last_hidden_state) {
+        const auto last_hidden_state_data = last_hidden_state.data<float>();
```

Comment:
as I see from API standpoint, `EmbeddingResult` can also store 8 bit types. I think you need to handle such case here, then